### DISCUSSION TITLE
Visual script property selector 4.0.alfa

### DISCUSSION THREAD

**Discussion by Gallilus** [godotengine/godot-proposals#4126] [Scripting]
Hi, wanted to start a discussion on the new visual script property selector for 4.0.alfa

### Why change it?
Selecting/ building/ configuring your node in visual script is not as simple as just knowing what you want. (copy past)
Even If you know what you want you may not find it, so it is also understanding where and how to find it.
So a simple contextual search resulted in many situations where people knew what they wanted but still did not find a satisfying result.

### Intended result.
Find more useful stuff.
Get more helpful data.
Bring the contextual part of searching to the foreground.

## Current status.
Find all stuff, maybe too much stuff.
- It now uses a method similar to Search Help.
- looking, true project classes.
- looking, VS nodes.
- (In the future I hope to add looking true project settings)

Find most helpful data. (if documentation exists)

Bring the contextual part of searching to the foreground.
- The default selections may need improvement and refinement
- It is still necessary to understand where and how to find objects.

### 3.4
![image](https://user-images.githubusercontent.com/20573784/155879624-e4af5c5b-a33c-419e-8579-59eca6e7cc11.png)
### 4.0.alfa
![image](https://user-images.githubusercontent.com/20573784/155879598-2a95c779-75d9-4bbb-809b-46a5d3f1afa2.png)

### Related Issues
https://github.com/godotengine/godot/issues/58069
https://github.com/godotengine/godot/issues/58128
https://github.com/godotengine/godot/issues/58129
https://github.com/godotengine/godot/issues/58118

**Comment by archerallstars** [comment]
I think the new node search window in Godot 4 makes coding in Visual Script slower. Let's assume you want to add an `array` related node. Here's what you would do in Godot 3 as shown in the screen recording below:

https://user-images.githubusercontent.com/1403194/157796931-9937c130-26a4-4e9b-b7c7-76fe0ef93452.mp4

And here's what you would do in Godot 4:

https://user-images.githubusercontent.com/1403194/157797079-e638440f-d0f2-428b-af53-d3cc5cc81cbe.mp4

As you can see, it's much slower to get what you want in Godot 4. This is due to the **search filter** that the user always needs to change to _search all_ first before they can do anything meaningful in Visual Script at all.

And not only that, you will have to enable the _class_ search to actually see the array nodes.

This implementation is the way to _prevent_ the users from finding the node they want!

The only improvement that I can see from the new search window is the filter bar. If the goal is to make the result nodes less clustered to the users, the new filter bar is the best tool for that, as shown in the screen recording below:

https://user-images.githubusercontent.com/1403194/157798322-ee7c3aed-0f09-498f-bf6a-f02a8db73596.mp4

IMO, the search filter box should be removed. And the class filter in the filter bar should be enabled by default.

**Reply by Gallilus** [reply]
The primary reason for the increased search complexity is that before it was sometimes impossible to find things.
even if you knew what you were looking for. With https://github.com/godotengine/godot/pull/59763 I hope that your click rate is reduced to a tolerable level?

**Comment by archerallstars** [comment]
Other than the issue caused by the new search filter box, the default turned-on/off items in the filter bar is also a problem. For example, if you try to add a node from the output port, all of the Visual Script nodes will be excluded by default (even though you're working in Visual Script), as shown in the screen recording below:

https://user-images.githubusercontent.com/1403194/158026923-b6de3337-f8da-4958-b69f-8544c863fa9c.mp4

And when compared to Godot 3:

https://user-images.githubusercontent.com/1403194/158026951-a09f2724-07ba-4dd3-91bf-5e94d1453665.mp4

Even `print`, which is the very typical node in debugging, is excluded from the search result by default. As you can see from the above screen recordings, adding the `print` node is much faster in Godot 3.

**Reply by Gallilus** [reply]
Indeed when you drag from a data port is assume you would like to do something with data.
While `print` is an understandable exception to the rule visual script nodes are provided by pulling from the white connection errors. `print` has to be connected anyway and then you do not even have to type anything.
